File: src/104.maximum-depth-of-binary-tree.cpp
/*
 * @lc app=leetcode id=104 lang=cpp
 *
 * [104] Maximum Depth of Binary Tree
 */
#include <algorithm>
using namespace std;
// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
// @lc code=start
class Solution {
 public:
  int maxDepth(TreeNode *root) {
    if (root == nullptr) {
      return 0;
    }

    int depth = 1;
    return repeat(root, depth);
  }

  int repeat(TreeNode *node, int depth) {
    if (node == nullptr) {
      return depth;
    }

    if (node->left || node->right) {
      depth++;
    }

    return max(repeat(node->left, depth), repeat(node->right, depth));
  }
};
// @lc code=end

File: Binary Tree/src/104.maximum-depth-of-binary-tree.js
/*
 * @lc app=leetcode id=104 lang=javascript
 *
 * [104] Maximum Depth of Binary Tree
 */


// Definition for a binary tree node.
class TreeNode {
  /**
   * @param {number} val
   * @param {TreeNode} left
   * @param {TreeNode} right
   */
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

// @lc code=start
/**
 * @param {TreeNode} node
 * @param {number} depth
 * @return {number}
 */
const repeat = (node, depth) => {
  if (node === null) {
    return depth;
  }

  return Math.max(repeat(node.left, depth + 1), repeat(node.right, depth + 1));
};

/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
  if (root === null) {
    return 0;
  }

  return repeat(root, 0);
};
// @lc code=end

